### PR TITLE
Reenable /analyze and fix all the warnings.

### PR DESCRIPTION
**Pull request by antkmsft**
Yes, we state that dereferencing null should never happen, but code analysis does not see that. It sees that we invoke `precondition_failed()` function, and that's it. So to a code analysis tool, after invoking that function, execution continues, and the code would go and dereference NULL pointer. Which, to us, is fine, but we have to explain it to a code analysis tool. `__analysis_assume` does that.

For the IoT samples, I **_had_** to rewrite some `if` statements (the ones that contain `AZ_SPAN_NULL`) to remove assignment from them. I tried a few less invasive methods (pragma warning disable, redefine AZ_SPAN_NULL as `(az_span) { 0 }` for MSVC), but it didn't work.
So modifying all the instances is the simplest fix (and maybe the only fix available; at least for now, because I do believe there's a bug in the analyze - both in the fact that it doesn't like our assignments in ifs() when AZ_SPAN_NULL is used, and that the warning can't be disabled).

If you want to give it another shot, now or later - feel free. I am not a fan of this solution as well. I like the `if(az_failed(rc = ...))` form. It is common for C to do this. It is unfortunate that `/analyze` does complain about it. What we do is clearly distinctive from the potential error when you write `if(i = get_int())` - in our case, we do pass a result of the assignment to a function that takes az_result, and it  returns bool, and that bool goes to the if().
`/analyze` did not find a bug. If you follow it's advice and replace `=` with `==` - you will actually make a bug, but the tool will be silent. And I hate we have to change our code not to make the code better, but to make the tool happy.

In the future, if it comes to that, it should be relatively easy to undo this: to find all instances, search for `if(az_failed(rc))`.

### CODE REVIEW

**Comment by ahsonkhan** [comment]
> For some reason, code analysis tool doesn't have an issue with us potentially dereferencing the NULL pointer when we build with precondition checking disabled.

That is certainly strange. We should dig into that a bit more. I would have expected the same `null referencing` warning to show up when preconditions are off. Maybe we are missing that OS/flags permutation in CI. Can you check locally?

File: sdk/inc/azure/core/internal/az_precondition_internal.h [hunk 1]
```diff
@@ -43,23 +43,30 @@
...
   } while (0)
 #endif // AZ_NO_PRECONDITION_CHECKING
 
-#define _az_PRECONDITION_RANGE(low, arg, max) _az_PRECONDITION((low <= arg && arg <= max))
+#define _az_PRECONDITION_RANGE(low, arg, max) _az_PRECONDITION(((low) <= (arg) && (arg) <= (max)))
 
-#define _az_PRECONDITION_NOT_NULL(arg) _az_PRECONDITION((arg != NULL))
-#define _az_PRECONDITION_IS_NULL(arg) _az_PRECONDITION((arg == NULL))
+#define _az_PRECONDITION_NOT_NULL(arg) _az_PRECONDITION(((arg) != NULL))
+#define _az_PRECONDITION_IS_NULL(arg) _az_PRECONDITION(((arg) == NULL))
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/internal/az_precondition_internal.h:69]: nit: What about other macros within the repo, including ones down below in this file?

nit: If there are more, can you extract this change out into a separate PR and keep this PR focused on just the null analysis issue?

**Comment by antkmsft** [comment]
> > For some reason, code analysis tool doesn't have an issue with us potentially dereferencing the NULL pointer when we build with precondition checking disabled.
> 
> That is certainly strange. We should dig into that a bit more. I would have expected the same `null referencing` warning to show up when preconditions are off. Maybe we are missing that OS/flags permutation in CI. Can you check locally?

No, it all looks fine, as you'd expect; and you would expect it to fail. It looks as we always use `/analyze`, regardless of whether we build with preconditions or not.

Another mystery - why disabling `C6011`, as we did previously, had no effect on this recent MSVC version? Which I am glad we are no longer doing, but it is still a mystery that looks like a compiler bug to me. Maybe one day our builds with precondition will also start failing all of a sudden, until we define precondition as `_az_ANALYSIS_ASSUME`.

File: eng/common/pipelines/templates/steps/create-pull-request.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@ parameters:
   PushArgs:
   WorkingDirectory: $(System.DefaultWorkingDirectory)
   PRTitle: not-specified
-  PRBody: not-specified 
+  PRBody: ''
```
Comment (ahsonkhan) [inline eng/common/pipelines/templates/steps/create-pull-request.yml:14]: I wouldn't have expected all these unrelated changes to come in. Bad merge?

Comment (antkmsft) [reply eng/common/pipelines/templates/steps/create-pull-request.yml]: yes, bad merge

**Comment by ahsonkhan** [comment]
@antkmsft - do you know if this will fix the iot sample warnings too: https://github.com/Azure/azure-sdk-for-c/issues/1137?

**Comment by antkmsft** [comment]
> @antkmsft - do you know if this will fix the iot sample warnings too: #1137?

I assume that if my PR build does pass, it fixes the issue? Because I reenable `/analyze`.

File: sdk/inc/azure/core/internal/az_precondition_internal.h [hunk 3]
```diff
@@ -43,12 +43,29 @@
 
 az_precondition_failed_fn az_precondition_failed_get_callback();
 
+// __analysis_assume() tells MSVC's code analysis tool about the assumptions we have, so it doesn't
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/internal/az_precondition_internal.h:46]: How are we telling it about our assumption for `null`?

File: sdk/inc/azure/core/internal/az_precondition_internal.h [hunk 4]
```diff
@@ -43,12 +43,29 @@
...
+// emit warnings for the statements that we put into _az_PRECONDITION().
+// Code analysis starts to fail with this condition some time around version 19.27; but
+// __analysis_assume() does appear at least as early as version 1920 (The first "Visual Studio
+// 2019"). Maybe it is available in 2017 - MSDN does have an article on __analysis_assume dated
+// 2017, but it looks like that version did not generate a warning in our code base (it likely
+// should; but it looks like it wasn't that fully featured prior to 2019). So, we do
+// __analysis_assume() for MSVCs starting with 2019. We don't want to go much earlier without
+// verifying, because if __analysis_assume() is not available on earlier compiler version, there
+// will be a compilation error.
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/internal/az_precondition_internal.h:55]: I think the blurb about 2017 and whether it had support or not is unnecessary noise. Remove those couple of sentences.

```suggestion
// Code analysis starts to fail with this condition some time around version 19.27; but
// __analysis_assume() does appear at least as early as version 1920 (The first "Visual Studio
// 2019"). So, we do
// __analysis_assume() for MSVCs starting with 2019. We don't want to go much earlier without
// verifying, because if __analysis_assume() is not available on earlier compiler version, there
// will be a compilation error.
```

Comment (ahsonkhan) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:55]: Maybe add a link to one of these:
https://docs.microsoft.com/en-us/cpp/code-quality/how-to-specify-additional-code-information-by-using-analysis-assume?view=vs-2019
See https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/using-the--analysis-assume-function-to-suppress-false-defects

File: sdk/src/azure/core/az_json_writer.c [hunk 5]
```diff
@@ -135,6 +135,11 @@ static AZ_NODISCARD bool _az_is_appending_value_valid(az_json_writer const* json
 }
 
 #ifndef AZ_NO_PRECONDITION_CHECKING
+// Declared to avoid warning C28278: Function appears with no prototype in scope.  Only limited
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_json_writer.c]: Interesting. Looks like these are needed because they are only defined/used within `AZ_NO_PRECONDITION_CHECKING` #if/def blocks.

File: sdk/inc/azure/core/_az_cfg.h [hunk 6]
```diff
@@ -25,8 +25,10 @@
 // automatic variable '...'
 #pragma warning(disable : 4221)
 
-// warning C6011: Dereferencing NULL pointer. Using _az_PRECONDITION_NOT_NULL
-#pragma warning(disable : 6011)
+// warning C28278 : Function appears with no prototype in scope. Only limited analysis can be
+// performed. Include the appropriate header or add a prototype. This warning also occurs if
+// parameter or return types are omitted in a function definition.
+#pragma warning(disable : 28278)
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/_az_cfg.h:31]: I think this is a good warning to keep enabled, and opt-out in the specific instances. If it is too noisy, then ignoring it is fine. Up to you.

File: sdk/src/azure/core/az_json_writer.c (hunk 5 above)
Comment (antkmsft) [reply sdk/src/azure/core/az_json_writer.c]: No, I tried to declare them first, and it didn't help. But I think, code analysis tool still has bugs, so it is unfixable with reasonable costs. Maybe, we can make it go away if we declare the function as non-static, but I don't think it's worth it. Eventually, code analysis will probably make its analysis tool smarter so even lack of forward declaration won't be a problem for it.

File: sdk/inc/azure/core/_az_cfg.h (hunk 6 above)
Comment (antkmsft) [reply sdk/inc/azure/core/_az_cfg.h:31]: My opinion is that the tool should be smart enough to figure out the signature itself. I did the minimum effort to declare the function before defining it, but that wasnt enough. I suspect if I not only declare it, but also remove static, it would make the warning to go away, but it is asking too much. This isn't about the problem with our code, it is the code analysis tool tells us that its current version is not smart enough to figure out all the info, and asing for help. I could put a couple of extra line, but I don't think we should change the design to compensate for code analysis current weaknesses.

File: sdk/src/azure/core/az_json_writer.c (hunk 5 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_writer.c]: > Maybe, we can make it go away if we declare the function as non-static, but I don't think it's worth it. 

Agreed.

>No, I tried to declare them first, and it didn't help. But I think, code analysis tool still has bugs, so it is unfixable with reasonable costs.

FYI, @barcharcraz

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 3 above)
Comment (antkmsft) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:46]: so we write `_az_PRECONDITION_NOT_NULL(ptr)`, which means `_az_PRECONDITION(ptr != NULL)`, which, among other things it expands into, does add `__analysis_assume(ptr != NULL)` <-- and this is what tells the code analysis about our assumption; nothing else does, and that's why it was emitting that warning.

**Comment by ahsonkhan** [comment]
Samples still failing:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=511583&view=logs&j=d0a3aede-fed8-508e-bfcc-69b7c0c9b6ed&t=68448d55-6b50-5f76-d0ee-4facc07a4ed2&l=119
```text
D:\a\1\s\sdk\samples\iot\paho_iot_hub_pnp_component_sample.c(511,1): error C2220: the following warning is treated as an error [D:\a\1\s\build\sdk\samples\iot\paho_iot_hub_pnp_component_sample.vcxproj]
<no file> : warning C6282: Incorrect operator:  assignment of constant in Boolean context. Consider using '==' instead. [D:\a\1\s\build\sdk\samples\iot\paho_iot_hub_pnp_component_sample.vcxproj]
<no file> : warning C6282: Incorrect operator:  assignment of constant in Boolean context. Consider using '==' instead. [D:\a\1\s\build\sdk\samples\iot\paho_iot_hub_pnp_component_sample.vcxproj]
  Generating Code...
  Building Custom Rule D:/a/1/s/sdk/samples/iot/CMakeLists.txt
  paho_iot_hub_pnp_sample.c
     Creating library D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_hub_pnp_sample.lib and object D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_hub_pnp_sample.exp
  paho_iot_hub_pnp_sample.vcxproj -> D:\a\1\s\build\sdk\samples\iot\Debug\paho_iot_hub_pnp_sample.exe
  Building Custom Rule D:/a/1/s/sdk/samples/iot/CMakeLists.txt
  paho_iot_hub_sas_telemetry_sample.c
D:\a\1\s\sdk\samples\iot\paho_iot_hub_sas_telemetry_sample.c(262,1): error C2220: the following warning is treated as an error [D:\a\1\s\build\sdk\samples\iot\paho_iot_hub_sas_telemetry_sample.vcxproj]
<no file> : warning C6282: Incorrect operator:  assignment of constant in Boolean context. Consider using '==' instead. [D:\a\1\s\build\sdk\samples\iot\paho_iot_hub_sas_telemetry_sample.vcxproj]
  Building Custom Rule D:/a/1/s/sdk/samples/iot/CMakeLists.txt
  paho_iot_hub_telemetry_sample.c
     Creating library D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_hub_telemetry_sample.lib and object D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_hub_telemetry_sample.exp
  paho_iot_hub_telemetry_sample.vcxproj -> D:\a\1\s\build\sdk\samples\iot\Debug\paho_iot_hub_telemetry_sample.exe
  Building Custom Rule D:/a/1/s/sdk/samples/iot/CMakeLists.txt
  paho_iot_hub_twin_sample.c
     Creating library D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_hub_twin_sample.lib and object D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_hub_twin_sample.exp
  paho_iot_hub_twin_sample.vcxproj -> D:\a\1\s\build\sdk\samples\iot\Debug\paho_iot_hub_twin_sample.exe
  Building Custom Rule D:/a/1/s/sdk/samples/iot/CMakeLists.txt
  paho_iot_provisioning_sample.c
     Creating library D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_provisioning_sample.lib and object D:/a/1/s/build/sdk/samples/iot/Debug/paho_iot_provisioning_sample.exp
  paho_iot_provisioning_sample.vcxproj -> D:\a\1\s\build\sdk\samples\iot\Debug\paho_iot_provisioning_sample.exe
  Building Custom Rule D:/a/1/s/sdk/samples/iot/CMakeLists.txt
  paho_iot_provisioning_sas_sample.c
D:\a\1\s\sdk\samples\iot\paho_iot_provisioning_sas_sample.c(432,1): error C2220: the following warning is treated as an error [D:\a\1\s\build\sdk\samples\iot\paho_iot_provisioning_sas_sample.vcxproj]
<no file> : warning C6282: Incorrect operator:  assignment of constant in Boolean context. Consider using '==' instead. [D:\a\1\s\build\sdk\samples\iot\paho_iot_provisioning_sas_sample.vcxproj]
##[error]Cmd.exe exited with code '1'.

```

**Comment by ahsonkhan** [comment]
Try changing the `rc = ...` assignment to be outside the `if(...)` block and see if that fixes it:

https://github.com/Azure/azure-sdk-for-c/blob/332102c172b7b3b2788e1c46a44e572e04d4673a/sdk/samples/iot/paho_iot_hub_sas_telemetry_sample.c#L222-L262

```C
static void generate_sas_key(void)
{
  az_result rc;

  // Create the POSIX expiration time from input minutes.
  uint64_t sas_duration = get_epoch_expiration_time_from_minutes(env_vars.sas_key_duration_minutes);

  // Get the signature that will later be signed.
  az_span sas_signature = AZ_SPAN_FROM_BUFFER(sas_signature_buffer);
  rc = rc = az_iot_hub_client_sas_get_signature(
      &hub_client, sas_duration, sas_signature, &sas_signature);
  if (az_failed(rc))
  {
    LOG_ERROR("Could not get the signature for SAS key: az_result return code 0x%08x.", rc);
    exit(rc);
  }

  // Generate the encoded, signed signature (b64 encoded, HMAC-SHA256 signing)
  az_span sas_encoded_signed_signature = AZ_SPAN_FROM_BUFFER(sas_encoded_signed_signature_buffer);
  sas_generate_encoded_signed_signature(
      env_vars.hub_sas_key,
      sas_signature,
      sas_encoded_signed_signature,
      &sas_encoded_signed_signature);

  // Get the resulting MQTT password, passing the base64 encoded, HMAC signed bytes
  size_t mqtt_password_length;
  rc = az_iot_hub_client_sas_get_password(
      &hub_client,
      sas_encoded_signed_signature,
      sas_duration,
      AZ_SPAN_NULL,
      mqtt_password_buffer,
      sizeof(mqtt_password_buffer),
      &mqtt_password_length);
  if (az_failed(rc))
  {
    LOG_ERROR("Could not get the password: az_result return code 0x%08x.", rc);
    exit(rc);
  }
}
```

**Comment by antkmsft** [comment]
Closes https://github.com/Azure/azure-sdk-for-c/issues/1137.

**Comment by antkmsft** [comment]
@ahsonkhan, @momuno - I updated IoT samples, see the updated PR description for details. There's a lot of changes, and I would appreciate if you could review that the replacements are correct (but it doesn't mean that there was lack of attention on my side).

**Comment by ahsonkhan** [comment]
> I do believe there's a bug in the analyze - both in the fact that it doesn't like our assignments in ifs(), and that the warning can't be disabled

I agree. @antkmsft, @barcharcraz - is the repo for the analyzer on GitHub? File an issue there with your example (also one for the null).

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 3 above)
Comment (ahsonkhan) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:46]: Gotcha. Isn't that an incorrect assumption to pass to `__analysis_assume`? `ptr` could certainly be `null` which is the whole point of the precondition.

If the condition was true, we wouldn't have `if (!condition) fail precondition`

File: sdk/samples/iot/paho_iot_hub_telemetry_sample.c [hunk 7]
```diff
@@ -147,11 +143,13 @@ static void connect_mqtt_client_to_iot_hub(void)
...
   // Connect MQTT client to the Azure IoT Hub.
-  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options);
+  if (rc != MQTTCLIENT_SUCCESS)
   {
     LOG_ERROR(
         "Failed to connect: MQTTClient return code %d.\n"
-        "If on Windows, confirm the AZ_IOT_DEVICE_X509_TRUST_PEM_FILE_PATH environment variable is "
+        "If on Windows, confirm the AZ_IOT_DEVICE_X509_TRUST_PEM_FILE_PATH environment variable "
+        "is "
```
Comment (ahsonkhan) [inline sdk/samples/iot/paho_iot_hub_telemetry_sample.c:152]: nit: Move `set correctly.` into this line.

File: sdk/samples/iot/sample_pnp_thermostat_component.c [hunk 8]
```diff
@@ -177,22 +176,23 @@ az_result pnp_thermostat_get_telemetry_message(
...
-  if (az_failed(
-          rc = pnp_get_telemetry_topic(
-              client,
-              NULL,
-              thermostat_component->component_name,
-              mqtt_message->topic,
-              mqtt_message->topic_length,
-              NULL)))
+
+  rc = pnp_get_telemetry_topic(
```
Comment (ahsonkhan) [inline sdk/samples/iot/sample_pnp_thermostat_component.c:180]: nit: merge the assignment  to the local declaration on line 178.

File: sdk/samples/iot/sample_pnp_thermostat_component.h [hunk 9]
```diff
@@ -9,6 +9,7 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#include <azure/core/az_json.h>
```
Comment (ahsonkhan) [inline sdk/samples/iot/sample_pnp_thermostat_component.h:12]: How did you detect that this needed to be added?

File: sdk/samples/iot/paho_iot_hub_pnp_component_sample.c [hunk 10]
```diff
@@ -997,45 +993,51 @@ static void property_callback(
...
+        LOG_AZ_SPAN("Payload:", mqtt_message.out_payload_span);
+
+        // Receive the response from the server.
+        mqtt_receive_message();
+      }
+      else
+      {
+        LOG_ERROR("Failed to update a property: az_result return code 0x%08x.", rc);
+        exit(rc);
+      }
```
Comment (ahsonkhan) [inline sdk/samples/iot/paho_iot_hub_pnp_component_sample.c:1040]: nit: Oh...this particular level of nesting is too deep. @momuno - try to refactor this so it doesn't need several if/else blocks.

**Comment by ahsonkhan** [comment]
I wonder why this coding pattern in `/src/` didn't get flagged. Do we run `/analyze` on the  source?

https://github.com/Azure/azure-sdk-for-c/blob/f4531ead209092413e05a1d7babd92c9d26273e2/sdk/src/azure/core/az_json_writer.c#L737-L741

File: sdk/samples/iot/sample_pnp_thermostat_component.h (hunk 9 above)
Comment (antkmsft) [reply sdk/samples/iot/sample_pnp_thermostat_component.h:12]: It does not compile, the type is unknown. You can see below, it is using json types, but never includes a json header.

File: sdk/inc/azure/core/internal/az_precondition_internal.h [hunk 11]
```diff
@@ -43,12 +43,28 @@
...
+#endif
+
 #ifdef AZ_NO_PRECONDITION_CHECKING
-#define _az_PRECONDITION(condition)
+#define _az_PRECONDITION(condition) _az_ANALYSIS_ASSUME(condition)
 #else
 #define _az_PRECONDITION(condition) \
   do \
   { \
+    _az_ANALYSIS_ASSUME(condition); \
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:67]: We do for all.

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 3 above)
Comment (antkmsft) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:46]: Yes, but we expect the precondition to never fail, which is what we tell via analysis_assume:
```
   precondition(ptr != NULL);
   *ptr = 3;
```

if  `precondition()` is defined simply as `if (!condition) { error_handler(); }`, then the code transforms to:
```
   if(ptr == NULL)
   {
      error_handler(); // where would execution go after error_handler() exits?
   }
   *ptr = 3; // it would go and dereference null
```
But now, it is:
```
   analysis_assume(ptr == NULL);
   if(ptr == NULL)
   {
      // We tell analysis that the execution will never go into this block.
      error_handler(); // where would execution go after error_handler() exits?
   }
   *ptr = 3; // since we told analysis that ptr can never be null, it is no longer concerned about dereferencing NULL
```

**Comment by antkmsft** [comment]
Update: I merged the latest changes from master. I found out that only rewriting the statements that contain `AZ_SPAN_NULL` is sufficient, so this is a smaller change. @ahsonkhan, the PR is ready to go. I think we should keep analyze_assume for all preconditions. Please approve faster if you can - it is costly to merge IoT hub changes as they are actively changed.

File: sdk/samples/iot/sample_pnp_thermostat_component.h (hunk 9 above)
Comment (ahsonkhan) [reply sdk/samples/iot/sample_pnp_thermostat_component.h]: In that case, how did it pass CI before and get merged to master?

https://github.com/Azure/azure-sdk-for-c/blob/f3a74661625adcf20c1df10ec0cb5e59247f9a9e/sdk/samples/iot/sample_pnp_thermostat_component.h#L4-L16

File: sdk/samples/iot/paho_iot_hub_pnp_component_sample.c (hunk 10 above)
Comment (momuno) [reply sdk/samples/iot/paho_iot_hub_pnp_component_sample.c]: I think I fixed this in my newest PR.  But just now seeing this comment!
